Ongoing code cleanup for Python 3 conversion

pl77 opened the pull request:
Just answering the "help wanted" badge on https://github.com/sabnzbd/sabnzbd/issues/1203.

I'll open this pull request to track my adjustments to the develop branch as I work through some of the functions and to prevent our work from overlapping.  

Please let me know if I'm breaking sabnzbd code conventions or you prefer me to focus on something else.  I'm doing my best not to change any of the intent of the underlying code, only clean it up for Python 3.5. 

Also, please let me know if Python 3.5 is still the target.  I could really clean up string formatting if I could use f-strings from 3.6!

thezoggy commented:
A lot of these changes break Sab. The T variable is used for translation (language) support.

Safihre commented on sabnzbd/filesystem.py:
```diff
@@ -792,23 +792,25 @@ def find_dir(p):
 else:
     try:
-        os.statvfs
+        # os.statvfs                # commented this because it does not appear to do anything.
```
What this does is just to see if `statvfs` is available, if not the `try` will catch it and we skip this function.

pl77 replied:
Thanks!  I was wondering about that! I will drop the # and add a comment to that effect.

pl77 replied:
I modified this to a straight function call to ensure Python actually calls it and doesn't ignore it with lazy-loading or whatever happens under the hood.  It's a simple call to the cwd now to make sure it's actually doing something.  It can always be reverted later if it causes a slowdown or other issues.

Safihre commented on sabnzbd/filesystem.py:
```diff
@@ -792,23 +792,25 @@ def find_dir(p):
-        pass
+        import win32api as w32
+    except ImportError:
+        w32 = None
```
Interesting approach, is this the recommended way?

pl77 replied:
I'm still trying to figure out the recommended way.  Normally an ImportError that isn't dealt with ends the whole script.  Since this keeps going I need to define the variable and catch it when the module isn't there.  

I changed the name just out of habit in case win32api was being called elsewhere to avoid clobbering it.  I mean, it shouldn't be called elsewhere, but I'm still working my way through this codebase and am very unfamiliar with it.  Until I'm more confident I understand it I was playing it safe.  Just like the "T" confusion I caused earlier!

Safihre commented on sabnzbd/filesystem.py:
```diff
@@ -300,16 +300,16 @@ def create_real_path(name, loc, path, umask=False, writable=True):
         my_dir = real_path(loc, path)
         if not os.path.exists(my_dir):
-            logging.info("%s directory: %s does not exist, try to create it", name, my_dir)
+            logging.info("{} directory: {} does not exist, try to create it".format(name, my_dir))
```
I enjoy the new `.format()`, but in the case of logging is it really more readable than using the parameters to the `info`/`debug`/etc functions?
That's also the way how we do it in all the other files :)

pl77 replied:
You are absolutely right, so I removed the 'format' from all logging functions.  As you stated elsewhere, I left it in for the string modifications.  It's a bit more robust when handling integers, etc., so there is some value in using it.

Safihre commented on sabnzbd/filesystem.py:
```diff
@@ -300,16 +300,16 @@ def create_real_path(name, loc, path, umask=False, writable=True):
-            logging.info("%s directory: %s does not exist, try to create it", name, my_dir)
+            logging.info("{} directory: {} does not exist, try to create it".format(name, my_dir))
             if not create_all_dirs(my_dir, umask):
-                logging.error(T("Cannot create directory %s"), clip_path(my_dir))
```
As @thezoggy mentioned, the `T` function is patched in `lang.py` so it's available everywhere as a more general way to translate instead of Python's weird `_` function. So we really need it!

pl77 replied:
> A lot of these changes break Sab. The T variable is used for translation (language) support.

Thanks @thezoggy and @Safihre, I will return the T variable back to its original position, lol.  I swear I did search and couldn't find a reference anywhere for it, but that makes way more sense.  

Ignore this update for now and I'll take your feedback into account and add a new commit that deals with this.

Safihre commented on sabnzbd/filesystem.py:
```diff
@@ -557,17 +557,17 @@ def get_unique_filename(path):
 
 @synchronized(DIR_LOCK)
-def recursive_listdir(dir):
+def recursive_listdir(target_dir):   # renamed to avoid shadowing out of scope variable
```
Great 💯

Safihre commented on sabnzbd/filesystem.py:
```diff
@@ -648,11 +648,11 @@ def get_filepath(path, nzo, filename):
             new_dirname = dirname
             if n:
-                new_dirname += "." + str(n)
+                new_dirname += ".{}".format(n)
```
Here I like the use of `format`!

Safihre commented on sabnzbd/utils/certgen.py:
```diff
@@ -57,17 +58,17 @@ def generate_local_cert(private_key, days_valid=3560, output_file="cert.cert", L
         if mylocalipv4:
             san_list.append(x509.IPAddress(ipaddress.IPv4Address(str(mylocalipv4))))
-    except:
+    except (crypto_exceptions.InternalError, ipaddress.AddressValueError, ipaddress.NetmaskValueError):
```
Are you sure these are all that can occur?

pl77 replied:
These were the ones that popped up when I ran a simple test, but you're absolutely right: there are likely more that can occur.  What I'll try and do is add some broader exceptions, but not leaving a bare exception, like BufferError or LookupErrors, that capture more without interfering with real bugs.

Safihre commented on sabnzbd/utils/checkdir.py:
```diff
@@ -1,4 +1,4 @@
-#!/usr/bin/env python
+#!/usr/bin/env python3
```
👍

Safihre commented on sabnzbd/utils/checkdir.py:
```diff
@@ -9,13 +9,16 @@
 
 debug = False
 
+# TODO: Rename function isFAT and variable FAT to PEP8-compliant 'is_fat32' and 'fat32. Check for any API impacts first'
```
Yes, we should do this 👍 
Just check for it's usages throughout the code.

Safihre commented on sabnzbd/utils/checkdir.py:
```diff
@@ -54,7 +57,7 @@ def isFAT(check_dir):
                 if result[4].startswith("FAT"):
                     FAT = True
-            except:
+            except (OSError, SystemError, IndexError, win32api.error):                  # Eliminate broad exception
```
Good change, the comment isn't needed.

Safihre commented on sabnzbd/utils/diskspeed.py:
```diff
@@ -50,22 +50,22 @@ def diskspeedmeasure(dirname):
     if len(sys.argv) >= 2:
-        dirname = sys.argv[1]
-        if not os.path.isdir(dirname):
+        directoryname = sys.argv[1]
```
👍

Safihre commented:
Thanks for the PR! I left some comments in the code.

While the CI/CD pipeline isn't super stable, it does give some feedback. For example we use the `black` code-formatter, so it's good to run that on all the listed files:
https://travis-ci.org/sabnzbd/sabnzbd/jobs/590686688#L421

pl77 commented:
> For example we use the `black` code-formatter, so it's good to run that on all the listed files:

Thanks again for the feedback, it was greatly appreciated.  I installed black and incorporated it into my PyCharm.  I also adjusted my line endings to 88 to match it's requirements and set the "T" builtin to be ignored in my own reviews.  These files seem to be passing tests, so I thought I'd do a quick commit to see if I'm on the right track and if the integration tools agree with me!  

Thanks again, this is a fantastic project and you folks are doing a great job.

Safihre commented:
So we actually use `-l120` (so line-length of 120) since otherwise it feels a bit too short and it makes these weird cvhanges.
Could you run `black` again with `-l120`? :)

pl77 commented:
> So we actually use `-l120` (so line-length of 120) since otherwise it feels a bit too short and it makes these weird changes.
> Could you run `black` again with `-l120`? :)

Oh thank goodness, 88 was really short!  I will rerun and recommit.  Thanks for your patience with me!

pl77 commented:
I'm going to close this mess and spend a bit of time taking a different approach.  It's getting a little unwieldy.

Safihre commented:
It was going pretty well :) looking forward to the next PR!

Have to say that I'm going on vacation to Africa Friday for 3 weeks, so it will be hard for me to review stuff then!

pl77 commented:
Don't worry, @Safihre, I'm still submitting my changes! 

After spending a few days working on it I feel a bit more comfortable with how the program is structured.  In some ways, it kind of felt like I had jumped into the middle of the lake and then tried to figure out which way I needed to swim.

In particular, I am spending a bit more time with the translation tooling scripts to make sure I don't irreversibly break something while I tweak, and to make sure the tests keep up with my changes.  I'm learning so much from this code base, it's a lot of fun!

My fork is still at the same link (https://github.com/pl77/sabnzbd), but I backed up my old commits to a local repo and started with a fresh clone.  I'm spending some time getting the testing suite fully operational on my dev machine before I commit any of my old code.  

**One question for you: when you do local development do you spin up docker instances to test new builds or just run the test suite in a local environment?** 

I'll put a note in my calendar to make sure a submit a new PR when you're back. I hope you enjoy your vacation!  It sounds amazing!  Africa is one of the countries I haven't visited yet and I'd love to some day.

thezoggy commented:
@pl77 If i recall correctly you just need to be in sab root dir when you execute it.
For building the translations you need to do:
https://github.com/sabnzbd/sabnzbd/blob/develop/INSTALL.txt#L92

pl77 commented:
Hahaha, you're totally right.  I had to set the proper location in PyCharm for the tests, as well as download a chromedriver.exe that matched my installed chrome, and add some environmental variables.  But now the tests are passing!   Thanks!

Safihre commented:
I personally created a Powershell script that just changes the directory to the root dir, sets the environment variables and then runs `pytest`. 👍

pl77 commented:
That's a good idea.  I was only trying to get them to run in PyCharm so I could run pause the tests at certain spots with the debugger.  Here's the configuration that ended up working for me, in case anyone reading this is struggling with their own IDE.  

Note the warning about no pytest runner in the selected interpreter.  I'm not sure why it's saying that, because pytest was installed in that environment and it had no problems running the tests... 

The test pane and menu for changing configurations:
![PyCharm pytest pane](https://user-images.githubusercontent.com/9884582/66064132-ce033700-e512-11e9-9c2d-f264d7cac564.png)

The configuration screen:
![PyCharm Configuration Screen](https://user-images.githubusercontent.com/9884582/66064133-ce033700-e512-11e9-8ed3-35437cd2c275.png)

Regardless, so far I've gotten "test_functional_config.py", "test_functional_misc.py", and "test_functional_downloads.py" to all pass on my local machine, which gives me much more confidence that I haven't broken anything significant when I go to submit my next PR.

pl77 commented:
One quick question: I noticed an assortment of files that had unix LF line endings and others with the windows CRLF line endings.  Do you have a preference as to which I standardize on?

Safihre commented:
> Do you have a preference as to which I standardize on?

Not really :)

thezoggy commented:
main thing is to make sure git is not set to do autocrlf correction, leave it as is.
we ran into this forever ago where devs on different os would keep checking all the line endings and producing unneeded changes.

pl77 commented:
Excellent, thanks guys.  @thezoggy that was the main reason I was asking; if you were all working on CRLF and I kept changing it to LF every time I committed I figured it'd drive you all nuts.  But it sounds like it's not a big deal regardless, so I won't make it one.

The only reason I was asking was in one type or another had been identified in the past as messing with any of the automated build systems.  I doubt it is a big problem, but I've been caught off guard in the past where running a CRLF file in a Linux environment caused all sorts of hassles that were difficult to debug.
